File: 0021_Merge_Two_Sorted_Lists/solution.cpp
#include <iostream>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};
class Solution {
    public:
        ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
            ListNode temp;
            ListNode* tail = &temp;
            while (list1 && list2) {
                if (list1->val <= list2->val) {
                    tail->next = list1;
                    list1 = list1->next;
                } else {
                    tail->next = list2;
                    list2 = list2->next;
                }
                tail = tail->next;
            }
            tail->next = list1 ? list1 : list2;
            return temp.next;
        }
};
void printList(ListNode* head) {
    while(head) {
        cout << head->val;
        if(head->next) {
            cout << " -> ";
        }
        head = head->next;
    }
    cout << endl;
}
ListNode* createNewList(const int arr[], int size) {
    if(size == 0) return nullptr;
    ListNode* head = new ListNode(arr[0]);
    ListNode* current = head;
    for(int i = 1; i < size; ++i) {
        current->next = new ListNode(arr[i]);
        current = current->next;
    }
    return head;
}
int main()  {
    Solution sol;
    int arr1[] = {1, 2, 4};
    int arr2[] = {1, 3, 4};
    ListNode* list1 = createNewList(arr1, sizeof(arr1)/sizeof(arr1[0]));
    ListNode* list2 = createNewList(arr2, sizeof(arr2)/sizeof(arr2[0]));
    ListNode* mergedList = sol.mergeTwoLists(list1, list2);
    cout << "Merged List: ";
    printList(mergedList);
    return 0;
}

File: 0021_Merge_Two_Sorted_Lists/solution.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def mergeTwoLists(self, list_A: ListNode, list_B: ListNode) -> ListNode:
        dummy = ListNode()
        tail = dummy
        while list_A and list_B:
            if list_A.val <= list_B.val:
                tail.next = list_A
                list_A = list_A.next
            else:
                tail.next = list_B
                list_B = list_B.next
            tail = tail.next
        tail.next = list_A if list_A else list_B
        return dummy.next
    def create_list(arr):
        if not arr:
            return None
        head = ListNode(arr[0])
        current = head
        for value in arr[1:]:
            current.next = ListNode(value)
            current = current.next
        return head
    def print_list(node):
        while node:
            print(node.val, end=" -> ")
            node = node.next
        print("None")
    def test():
        sol = Solution()
        list_A = sol.create_list([1, 2, 4])
        list_B = sol.create_list([1, 3, 4])
        merged_list = sol.mergeTwoLists(list_A, list_B)
        sol.print_list(merged_list)
    if __name__ == "__main__":
        test()